KOGITO-4520: Design and adopt a strategy for auto/manual resize the Table columns in the Boxed Expression Editor

karreiro opened the pull request:
[KOGITO-4520](https://issues.redhat.com/browse/KOGITO-4520): Design and adopt a strategy for auto/manual resize the Table columns in the Boxed Expression Editor

Here's the live demo: [karreiro.com/kogito-editors-java](https://karreiro.com/kogito-editors-java/)

![2021-05-17 13 57 12](https://user-images.githubusercontent.com/1079279/118484954-1a1ae380-b718-11eb-8144-07a516d30505.gif)

This PR doesn't aim to cover [KOGITO-4587](https://issues.redhat.com/browse/KOGITO-4587) and [KOGITO-5131](https://issues.redhat.com/browse/KOGITO-5131).

karreiro committed:
KOGITO-4520: Design and adopt a strategy for auto/manual resize the Table columns in the Boxed Expression Editor

jomarko commented:
In the area of clearing expression, the cursor gets unexpected icon, it is a cross, or a plus icon, I am sorry, however I am not able to capture it on screenshot.

jomarko commented:
### Manual verification checklist
Add all possible types of expressions (it means so many rows all types are present). Consequently try to change width of each expression. Try to add/remove column (where applicable) in nested expression
- [x] context as top level expression
- [x] list as top level expression 
- [x] invocation as top level expression
- [x] function as top level expression

Sanity check for browsers
- [x] chrome
- [x] firefox

jomarko commented:
@karreiro thank you, I didn't find an issue, but https://github.com/karreiro/kogito-editors-java/pull/1#issuecomment-843190995 do you agree with reporting it? is that something we can improve in future?

karreiro committed:
Use the default cursor into cells border

karreiro commented:
Thank you for the review, @jomarko! I agree. I just fixed it in a new commit, and now we have the default cursor in the border of the cells:

![2021-05-19 09 15 56](https://user-images.githubusercontent.com/1079279/118771500-e90af200-b882-11eb-9dd5-4c72c1f42dda.gif)

vpellegrino commented:
I agree with @jomarko that behaviour looks great.
Just need to finish implementation review by this afternoon. 
Thanks @karreiro

vpellegrino commented on kie-wb-common-react/boxed-expression-editor/src/api/Table.ts:
```diff
@@ -105,6 +105,8 @@ export interface Column {
   /** Column width */
   width?: string | number;
+  /** Set column width */
+  setWidth?: (width: string | number) => void;
```
which is the difference between `width` property and `setWidth` function? In which context should we use one or another? Please explain it in the documentation

karreiro replied:
This method provides a bit more flexibility for `Table.ts` API users. Thus, they can rely on the `width` property, and they can also rely on a `setWidth` function (possibly based on `useState` hook) to set the component width. The `InvocationExpression` benefits itself of this API :-)

vpellegrino commented on kie-wb-common-react/boxed-expression-editor/src/__tests__/components/ContextExpression/ContextExpression.test.tsx:
```diff
@@ -109,3 +104,26 @@ describe("ContextExpression tests", () => {
+    ...actualReact,
+    useContext: useContext,
+  };
+});
```
we are repeating the same mocking in multiple tests, isn't possible to move this code in a common function or in the test-utils section?

karreiro replied:
Fixed :-) Thanks!

vpellegrino commented on kie-wb-common-react/boxed-expression-editor/src/components/BoxedExpressionEditor/BoxedExpressionEditor.tsx:
```diff
@@ -39,22 +40,35 @@ const BoxedExpressionEditor: (props: BoxedExpressionEditorProps) => JSX.Element
-    >
-      <BoxedExpressionGlobalContext.Provider
-        value={{ boxedExpressionEditorRef, currentlyOpenedHandlerCallback, setCurrentlyOpenedHandlerCallback }}
+  return useMemo(
```
just for my curiosity: why need to use `useMemo` there? performance improvement?

karreiro replied:
Yes, thus we avoid re-renders due to `supervisorHash` state :-)

vpellegrino commented on kie-wb-common-react/boxed-expression-editor/src/components/ContextExpression/ContextExpression.tsx:
```diff
@@ -164,21 +165,33 @@ export const ContextExpression: React.FunctionComponent<ContextProps> = ({
         onRowAdding={onRowAdding}
         onRowsUpdate={setRows}
-        handlerConfiguration={noHandlerMenu ? undefined : getHandlerConfiguration(i18n, i18n.contextEntry)}
+        handlerConfiguration={getHandlerConfiguration(i18n, i18n.contextEntry)}
```
why are you removing this prop? It is useful when context expression is nested in the function expression

karreiro replied:
Fixed :-) Thanks!

vpellegrino commented on kie-wb-common-react/boxed-expression-editor/src/components/InvocationExpression/InvocationExpression.tsx:
```diff
@@ -83,39 +83,44 @@ export const InvocationExpression: React.FunctionComponent<InvocationProps> = ({
-        onBlur={onFunctionDefinitionBlur}
-        defaultValue={functionDefinition.current}
+        defaultValue={functionName}
+        onBlur={(e) => setFunctionName(e.target.value)}
```
IMHO, this function should be hold in a `useCallback`

karreiro replied:
Fixed :-) Thanks!

vpellegrino commented on kie-wb-common-react/boxed-expression-editor/src/components/Resizer/ResizerSupervisor.tsx:
```diff
@@ -0,0 +1,36 @@
+    return () => clearTimeout(id);
+  }, [supervisorHash]);
+
+  return <div>{children}</div>;
```
why do we need this div? if you are using it for HTML structure purpose, I would assign to it a className, otherwise you can just use `React.Fragment` to have an empty enclosing tag

karreiro replied:
Fixed :-) Thanks!

vpellegrino commented on kie-wb-common-react/boxed-expression-editor/src/components/Resizer/dom/DOMSession.ts:
```diff
@@ -0,0 +1,62 @@
+
+import { Cell } from "./";
+
+export const CELL_CSS_SELCTOR = ".react-resizable";
```
typo :)

karreiro replied:
Fixed :-) Thanks!

vpellegrino reviewed (commented):
Overall solution looks good! Just few comments

karreiro committed:
- DRY tests
- Fix typo on 'CELL_CSS_SELCTOR' (~> CELL_CSS_SELECTOR)
- Remove useless HTMLDivElement from ResizerSupervisor.tsx
- Add useCallback into setFunctionName function (InvocationExpression.tsx)
- Bring noHandlerMenu back (ContextExpression.tsx)

karreiro commented:
Thank you for the review, @vpellegrino :-) Changes applied!

vpellegrino reviewed (approved):
thanks @karreiro
